pkg/web/interceptors.go

```diff
@@ -0,0 +1,137 @@
+package web
+
+import (
+	"github.com/Peripli/service-manager/pkg/extension"
+	"github.com/Peripli/service-manager/pkg/log"
+	"github.com/Peripli/service-manager/pkg/types"
+)
+
+type concreteBuilder interface {
+	Apply(interceptables []extension.Interceptable, interceptorType types.ObjectType, orderer extension.Ordered)
+}
+
+type interceptorBuilder struct {
+	concreteBuilder
+
+	orderer         *extension.OrderedProviderImpl
+	interceptorType types.ObjectType
+	interceptables  []extension.Interceptable
+}
+
+func (creator *interceptorBuilder) Apply() {
+	if creator.orderer == nil {
+		creator.orderer = &extension.OrderedProviderImpl{}
+	}
+	creator.concreteBuilder.Apply(creator.interceptables, creator.interceptorType, creator.orderer)
+}
+
+func (creator *interceptorBuilder) Before(name string) *interceptorBuilder {
+	return creator.TxBefore(name).APIBefore(name)
+}
+
+func (creator *interceptorBuilder) After(name string) *interceptorBuilder {
+	return creator.TxAfter(name).APIAfter(name)
+}
+
+func (creator *interceptorBuilder) TxBefore(name string) *interceptorBuilder {
```

Comment:
Currently we chain the interceptors as if they were filters which may not be ideal. We dont really have a use case where we want one hook to have a pair of (pretx,posttx), (preapi,postapi). And saying TxBefore may actually mean that you have a postTx logic that you want to execute after the postTx of the interceptor with `name`.

We currently have a usecase in which we have a broker hook (onPreApi -> get catalog) (onPostTx -> store catalog) and entitlements hook (onPre -> get entitlements) and (onPostTx -> store visibilities)

Semantically speaking all we want is that the broker hook stuff is executed before the entitltments hook stuff (we want to get the catalog first, then get the entitlements.. then first we want to store the catalog and then store the  visibilities).  so we register the entitlements Hook OnAPIBefore the broker hook. But because storing of catalog and visibilities is done PostTx and PostTx logic is reversed when executed we now have to register the entitlements hook TxAfter the broker hook even though what we mean is actually (semantically) TxBefore because we want to store the catalog first (broker hook) and the visibilities afterwards (entitlements hook).

This is way too confusing I think and we should try to instead of wrapping one into the other go for linear loop chainning. Then semantically and go-api-wise you would use the method that would "do what it says"